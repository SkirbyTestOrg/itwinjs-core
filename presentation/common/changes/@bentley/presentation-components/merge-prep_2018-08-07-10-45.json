{
  "changes": [
    {
      "packageName": "@bentley/presentation-components",
      "comment": "Rename `ECPresentation` to `Presentation`",
      "type": "minor"
    }
  ],
  "packageName": "@bentley/presentation-components",
  "email": "grigas.petraitis@bentley.com"
}