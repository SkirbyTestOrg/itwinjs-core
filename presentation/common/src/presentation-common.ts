/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

export {
  ClassId, InstanceId, InstanceKey,
  ClassInfo, EnumerationChoice, EnumerationInfo, KindOfQuantityInfo,
  PropertyInfo, RelatedClassInfo, RelationshipPath,
} from "./EC";
export { PresentationError, PresentationStatus } from "./Error";
export { KeySet, Keys, Key } from "./KeySet";
export { PersistentKeysContainer } from "./PersistentKeysContainer";
export {
  RequestOptions, HierarchyRequestOptions, ContentRequestOptions,
  LabelRequestOptions, SelectionScopeRequestOptions,
  PageOptions, Paged, RequestOptionsWithRuleset,
} from "./PresentationManagerOptions";
export {
  PresentationRpcInterface, PresentationRpcRequestOptions,
  LabelRpcRequestOptions, ClientStateSyncRequestOptions, ContentRpcRequestOptions,
  HierarchyRpcRequestOptions, SelectionScopeRpcRequestOptions,
  PresentationRpcResponse,
} from "./PresentationRpcInterface";
export { RpcRequestsHandler, RpcRequestsHandlerProps, IClientStateHolder } from "./RpcRequestsHandler";
export { RulesetVariablesState, VariableValueTypes, VariableValue } from "./RulesetVariables";
export { RegisteredRuleset, RulesetManagerState } from "./RegisteredRuleset";
export { RulesetsFactory } from "./RulesetsFactory";
export { LoggingNamespaces } from "./Logging";
export {
  Omit, Subtract, ValuesDictionary,
  getInstancesCount,
} from "./Utils";

/** @module UnifiedSelection */
export { SelectionScope } from "./selection/SelectionScope";

/** @module Content */
export { CategoryDescription } from "./content/Category";
export { Content } from "./content/Content";
export {
  Descriptor, DescriptorSource, DescriptorOverrides,
  SelectClassInfo, SelectionInfo, SortDirection, ContentFlags,
} from "./content/Descriptor";
export { DefaultContentDisplayTypes } from "./content/DisplayTypes";
export { EditorDescription } from "./content/Editor";
export { Field, PropertiesField, NestedContentField } from "./content/Fields";
export { Item } from "./content/Item";
export { Property } from "./content/Property";
export {
  PropertyValueFormat, TypeDescription, PrimitiveTypeDescription,
  ArrayTypeDescription, StructTypeDescription, StructFieldMemberDescription,
  BaseTypeDescription,
} from "./content/TypeDescription";
export {
  Value, ValuesArray, ValuesMap,
  DisplayValue, DisplayValuesArray, DisplayValuesMap,
  NestedContentValue,
} from "./content/Value";

/** @module Hierarchies */
export { NodeKey, NodeKeyPath, StandardNodeTypes } from "./hierarchy/Key";
export {
  BaseNodeKey, ECInstanceNodeKey, GroupingNodeKey,
  ECClassGroupingNodeKey, ECPropertyGroupingNodeKey, LabelGroupingNodeKey,
} from "./hierarchy/Key";
export { Node } from "./hierarchy/Node";
export { NodePathElement, NodePathFilteringData } from "./hierarchy/NodePathElement";

/** @module PresentationRules */
export { NavigationRule, NavigationRuleBase } from "./rules/hierarchy/NavigationRule";
export { RootNodeRule } from "./rules/hierarchy/RootNodeRule";
export { ChildNodeRule } from "./rules/hierarchy/ChildNodeRule";
export {
  ChildNodeSpecification, ChildNodeSpecificationTypes,
  ChildNodeSpecificationBase, DefaultGroupingPropertiesContainer,
} from "./rules/hierarchy/ChildNodeSpecification";
export { AllInstanceNodesSpecification } from "./rules/hierarchy/AllInstanceNodesSpecification";
export { AllRelatedInstanceNodesSpecification } from "./rules/hierarchy/AllRelatedInstanceNodesSpecification";
export { RelatedInstanceNodesSpecification } from "./rules/hierarchy/RelatedInstanceNodesSpecification";
export { InstanceNodesOfSpecificClassesSpecification } from "./rules/hierarchy/InstanceNodesOfSpecificClassesSpecification";
export {
  CustomQueryInstanceNodesSpecification, QuerySpecification, QuerySpecificationBase,
  QuerySpecificationTypes, StringQuerySpecification, ECPropertyValueQuerySpecification,
} from "./rules/hierarchy/CustomQueryInstanceNodesSpecification";
export { CustomNodeSpecification } from "./rules/hierarchy/CustomNodeSpecification";
export { SubCondition } from "./rules/hierarchy/SubCondition";

export { CustomizationRule } from "./rules/customization/CustomizationRule";
export { CheckBoxRule } from "./rules/customization/CheckBoxRule";
export { ImageIdOverride } from "./rules/customization/ImageIdOverride";
export {
  InstanceLabelOverride, InstanceLabelOverrideValueSpecificationType, InstanceLabelOverrideValueSpecificationBase,
  InstanceLabelOverrideValueSpecification, InstanceLabelOverridePropertyValueSpecification,
  InstanceLabelOverrideBriefcaseIdSpecification, InstanceLabelOverrideClassLabelSpecification,
  InstanceLabelOverrideClassNameSpecification, InstanceLabelOverrideLocalIdSpecification,
  InstanceLabelOverrideCompositeValueSpecification,
} from "./rules/customization/InstanceLabelOverride";
export { LabelOverride } from "./rules/customization/LabelOverride";
export { SortingRule, PropertySortingRule, DisabledSortingRule, SortingRuleBase } from "./rules/customization/SortingRule";
export { StyleOverride } from "./rules/customization/StyleOverride";
export {
  GroupingRule, GroupingSpecification, GroupingSpecificationTypes,
  SameLabelInstanceGroup, ClassGroup,
  PropertyGroup, PropertyGroupingValue, PropertyRangeGroupSpecification,
  GroupingSpecificationBase,
} from "./rules/customization/GroupingRule";

export { ContentRule } from "./rules/content/ContentRule";
export { ContentSpecification, ContentSpecificationTypes, ContentSpecificationBase } from "./rules/content/ContentSpecification";
export { ContentInstancesOfSpecificClassesSpecification } from "./rules/content/ContentInstancesOfSpecificClassesSpecification";
export { ContentRelatedInstancesSpecification } from "./rules/content/ContentRelatedInstancesSpecification";
export { SelectedNodeInstancesSpecification } from "./rules/content/SelectedNodeInstancesSpecification";

export { ContentModifier } from "./rules/content/modifiers/ContentModifier";
export {
  RelatedPropertiesSpecification, RelationshipMeaning,
  RelatedPropertiesSpecialValues,
} from "./rules/content/modifiers/RelatedPropertiesSpecification";
export { CalculatedPropertiesSpecification } from "./rules/content/modifiers/CalculatedPropertiesSpecification";
export { PropertiesDisplaySpecification } from "./rules/content/modifiers/PropertiesDisplaySpecification";
export {
  PropertyEditorParameters, PropertyEditorParametersBase,
  PropertyEditorsSpecification, PropertyEditorParameterTypes,
  PropertyEditorJsonParameters, PropertyEditorMultilineParameters,
  PropertyEditorRangeParameters, PropertyEditorSliderParameters,
} from "./rules/content/modifiers/PropertyEditorsSpecification";

export { Ruleset, SupplementationInfo } from "./rules/Ruleset";
export { Rule, RuleTypes, RuleBase, ConditionContainer } from "./rules/Rule";
export { VariablesGroup, Variable, VariableValueType } from "./rules/Variables";
export { RelatedInstanceSpecification } from "./rules/RelatedInstanceSpecification";
export { RelationshipDirection } from "./rules/RelationshipDirection";
export { SingleSchemaClassSpecification, MultiSchemaClassesSpecification } from "./rules/ClassSpecifications";
export { SchemasSpecification } from "./rules/SchemasSpecification";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
/* istanbul ignore next */
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("presentation-common", BUILD_SEMVER);
}
