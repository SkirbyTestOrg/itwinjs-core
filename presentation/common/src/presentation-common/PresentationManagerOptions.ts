/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import { Ruleset } from "./rules/Ruleset";
import { RulesetVariable } from "./RulesetVariables";

/**
 * Enumeration of standard request priorities.
 * @public
 */
export enum RequestPriority {
  /** Priority for pre-loading requests */
  Preload = 0,

  /** Priority for general requests */
  Normal = 1000,

  /** Max possible priority */
  Max = Number.MAX_SAFE_INTEGER,
}

/** @alpha */
export enum PresentationUnitSystem {
  Metric = "metric",
  BritishImperial = "british-imperial",
  UsCustomary = "us-customary",
  UsSurvey = "us-survey",
}

/**
 * A generic request options type used for both hierarchy and content requests
 * @public
 */
export interface RequestOptions<TIModel> {
  /** iModel to request data from */
  imodel: TIModel;

  /** Optional locale to use when formatting / localizing data */
  locale?: string;

  /**
   * Optional request priority. Higher priority requests are handled first.
   * Defaults to [[RequestPriority.Normal]]
   */
  priority?: number;
}

/**
 * Options for requests that require presentation ruleset. Not
 * meant to be used directly, see one of the subclasses.
 *
 * @public
 */
export interface RequestOptionsWithRuleset<TIModel> extends RequestOptions<TIModel> {
  /** Ruleset or id of the ruleset to use when requesting data */
  rulesetOrId: Ruleset | string;

  /** Ruleset variables to use when requesting data */
  rulesetVariables?: RulesetVariable[];
}

/**
 * Request type for hierarchy requests
 * @public
 */
export interface HierarchyRequestOptions<TIModel> extends RequestOptionsWithRuleset<TIModel> { }

/**
 * Request type for content requests
 * @public
 */
export interface ContentRequestOptions<TIModel> extends RequestOptionsWithRuleset<TIModel> {
  /**
   * Unit system to use when formatting property values with units. Default presentation
   * unit is used if unit system is not specified.
   *
   * @alpha
   */
  unitSystem?: PresentationUnitSystem;
}

/**
 * Request type for label requests
 * @public
 */
export interface LabelRequestOptions<TIModel> extends RequestOptions<TIModel> { }

/**
 * Request options used for selection scope related requests
 * @public
 */
export interface SelectionScopeRequestOptions<TIModel> extends RequestOptions<TIModel> { }

/**
 * Data structure for comparing presentation data after ruleset or ruleset variable changes.
 * @alpha
 */
export interface PresentationDataCompareOptions<TIModel> extends RequestOptionsWithRuleset<TIModel> {
  prev: {
    rulesetOrId: Ruleset | string;
  } | {
    rulesetVariables: RulesetVariable[];
  };
}

/**
 * Paging options
 * @public
 */
export interface PageOptions {
  /** Inclusive start 0-based index of the page */
  start?: number;
  /** Maximum size of the page */
  size?: number;
}

/**
 * A wrapper type that injects [[PageOptions]] into supplied type
 * @public
 */
export type Paged<TOptions extends {}> = TOptions & {
  /** Optional paging parameters */
  paging?: PageOptions;
};
