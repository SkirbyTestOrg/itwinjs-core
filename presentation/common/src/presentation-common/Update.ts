/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import { Node } from "./hierarchy/Node";

/** @alpha */
export const UPDATE_FULL = "FULL";

/** @alpha */
export interface UpdateInfo {
  [rulesetId: string]: {
    hierarchy?: HierarchyUpdateInfo;
    content?: ContentUpdateInfo;
  };
}

/** @alpha */
export type HierarchyUpdateInfo = typeof UPDATE_FULL | Array<NodeInsertionInfo | NodeDeletionInfo | NodeUpdateInfo>;

/** @alpha */
export type ContentUpdateInfo = typeof UPDATE_FULL;

/** @alpha */
export interface NodeInsertionInfo {
  type: "Insert";
  position: number;
  node: Node;
}

/** @alpha */
export interface NodeDeletionInfo {
  type: "Delete";
  node: Node;
}

/** @alpha */
export interface NodeUpdateInfo {
  type: "Update";
  node: Node;
  changes: Array<{
    name: string;
    old: unknown;
    new: unknown;
  }>;
}
