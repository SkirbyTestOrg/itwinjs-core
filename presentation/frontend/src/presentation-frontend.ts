/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

export { default as PersistenceHelper } from "./PersistenceHelper";
export { default as Presentation } from "./Presentation";
export { default as PresentationManager } from "./PresentationManager";
export { default as RulesetVariablesManager } from "./RulesetVariablesManager";

/** @module UnifiedSelection */
export { default as SelectionChangeEvent, SelectionChangeEventArgs, SelectionChangeType, SelectionChangesListener } from "./selection/SelectionChangeEvent";
export { default as ISelectionProvider } from "./selection/ISelectionProvider";
export { default as SelectionManager } from "./selection/SelectionManager";
export { default as SelectionHandler } from "./selection/SelectionHandler";
