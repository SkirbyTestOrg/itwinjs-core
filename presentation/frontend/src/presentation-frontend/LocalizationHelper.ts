/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { Node, LabelCompositeValue, LabelDefinition, Content, Item } from "@bentley/presentation-common";
import { Presentation } from "./Presentation";

const NAMESPACES = ["BisCore", "ECPresentation", "RulesEngine"];

/** @internal */
export class LocalizationHelper {
  public static async registerNamespaces() {
    const localizationPromises = NAMESPACES.map(async (namespace) => Presentation.i18n.registerNamespace(namespace).readFinished);
    await Promise.all(localizationPromises);
  }

  public static unregisterNamespaces() {
    NAMESPACES.map((namespace) => Presentation.i18n.unregisterNamespace(namespace));
  }

  public translate(stringId: string) {
    const key = stringId.replace(/^@|@$/g, "");
    return Presentation.i18n.translate(key, { defaultValue: stringId });
  }

  public getLocalizedNodes(nodes: Node[]): Node[] {
    for (const node of nodes)
      this.translateNode(node);
    return nodes;
  }

  public getLocalizedLabelDefinition(labelDefinition: LabelDefinition): LabelDefinition {
    this.translateLabelDefinition(labelDefinition);
    return labelDefinition;
  }

  public getLocalizedLabelDefinitions(labelDefinitions: LabelDefinition[]) {
    labelDefinitions.forEach((labelDefinition) => this.translateLabelDefinition(labelDefinition));
    return labelDefinitions;
  }

  public getLocalizedContent(content: Content | undefined) {
    if (content !== undefined)
      content.contentSet.forEach((item) => this.translateContentItem(item));
    return content;
  }

  private translateContentItem(item: Item) {
    this.translateLabelDefinition(item.label);
  }

  private translateNode(node: Node) {
    this.translateLabelDefinition(node.label);
  }

  private translateLabelDefinition(labelDefinition: LabelDefinition) {
    const translateComposite = (compositeValue: LabelCompositeValue) => {
      compositeValue.values.map((value) => this.translateLabelDefinition(value));
    };

    if (labelDefinition.typeName === "composite")
      translateComposite(labelDefinition.rawValue as LabelCompositeValue);
    else if (labelDefinition.typeName === "string")
      labelDefinition.rawValue = this.translate(labelDefinition.rawValue as string);
  }

}
