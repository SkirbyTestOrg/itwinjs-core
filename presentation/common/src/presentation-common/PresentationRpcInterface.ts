/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module RPC
 */

import { Id64String } from "@bentley/bentleyjs-core";
import { IModelRpcProps, RpcInterface } from "@bentley/imodeljs-common";
import { ContentJSON } from "./content/Content";
import { DescriptorJSON, DescriptorOverrides, SelectionInfo } from "./content/Descriptor";
import { InstanceKeyJSON } from "./EC";
import { PresentationStatus } from "./Error";
import { NodeKeyJSON } from "./hierarchy/Key";
import { NodeJSON } from "./hierarchy/Node";
import { NodePathElementJSON } from "./hierarchy/NodePathElement";
import { KeySetJSON } from "./KeySet";
import { LabelDefinitionJSON } from "./LabelDefinition";
import {
  ContentRequestOptions, HierarchyRequestOptions, LabelRequestOptions, Paged, PresentationDataCompareOptions, SelectionScopeRequestOptions,
} from "./PresentationManagerOptions";
import { SelectionScope } from "./selection/SelectionScope";
import { PartialHierarchyModificationJSON } from "./Update";
import { Omit } from "./Utils";

/**
 * Base options for all presentation RPC requests.
 * @public
 */
export type PresentationRpcRequestOptions<TManagerRequestOptions> = Omit<TManagerRequestOptions, "imodel"> & {
  /** ID of the client requesting data */
  clientId?: string;
};

/**
 * Data structure for presentation RPC responses
 * @public
 */
export type PresentationRpcResponse<TResult = undefined> = Promise<{
  /** Response status code */
  statusCode: PresentationStatus;
  /** In case of an error response, the error message */
  errorMessage?: string;
  /** In case of a success response, the result */
  result?: TResult;
}>;

/**
 * Data structure for hierarchy request options.
 * @public
 */
export type HierarchyRpcRequestOptions = PresentationRpcRequestOptions<HierarchyRequestOptions<any>>;
/**
 * Data structure for content request options.
 * @public
 */
export type ContentRpcRequestOptions = PresentationRpcRequestOptions<ContentRequestOptions<any>>;
/**
 * Data structure for label request options.
 * @public
 */
export type LabelRpcRequestOptions = PresentationRpcRequestOptions<LabelRequestOptions<any>>;
/**
 * Data structure for selection scope request options.
 * @public
 */
export type SelectionScopeRpcRequestOptions = PresentationRpcRequestOptions<SelectionScopeRequestOptions<any>>;
/**
 * Data structure for ruleset variable request options.
 * @public
 */
export type RulesetVariableRpcRequestOptions = PresentationRpcRequestOptions<{ rulesetId: string }>;
/**
 * Data structure for comparing presentation data after ruleset or ruleset variable changes.
 * @alpha
 */
export type PresentationDataCompareRpcOptions = PresentationRpcRequestOptions<PresentationDataCompareOptions<any>>;

/**
 * Interface used for communication between Presentation backend and frontend.
 *
 * @public
 */
export class PresentationRpcInterface extends RpcInterface {
  /** The immutable name of the interface. */
  public static readonly interfaceName = "PresentationRpcInterface"; // tslint:disable-line: naming-convention

  /** The semantic version of the interface. */
  public static interfaceVersion = "2.1.0";

  /*===========================================================================================
    NOTE: Any add/remove/change to the methods below requires an update of the interface version.
    NOTE: Please consult the README in core/common/src/rpc for the semantic versioning rules.
  ===========================================================================================*/

  public async getNodesAndCount(_token: IModelRpcProps, _options: Paged<HierarchyRpcRequestOptions>, _parentKey?: NodeKeyJSON): PresentationRpcResponse<{ nodes: NodeJSON[], count: number }> { return this.forward(arguments); }
  public async getNodes(_token: IModelRpcProps, _options: Paged<HierarchyRpcRequestOptions>, _parentKey?: NodeKeyJSON): PresentationRpcResponse<NodeJSON[]> { return this.forward(arguments); }
  public async getNodesCount(_token: IModelRpcProps, _options: HierarchyRpcRequestOptions, _parentKey?: NodeKeyJSON): PresentationRpcResponse<number> { return this.forward(arguments); }
  public async getNodePaths(_token: IModelRpcProps, _options: HierarchyRpcRequestOptions, _paths: InstanceKeyJSON[][], _markedIndex: number): PresentationRpcResponse<NodePathElementJSON[]> { return this.forward(arguments); }
  public async getFilteredNodePaths(_token: IModelRpcProps, _options: HierarchyRpcRequestOptions, _filterText: string): PresentationRpcResponse<NodePathElementJSON[]> { return this.forward(arguments); }
  /** @alpha Hierarchy loading performance needs to be improved before this becomes publicly available. */
  public async loadHierarchy(_token: IModelRpcProps, _options: HierarchyRpcRequestOptions): PresentationRpcResponse<void> { return this.forward(arguments); }

  public async getContentDescriptor(_token: IModelRpcProps, _options: ContentRpcRequestOptions, _displayType: string, _keys: KeySetJSON, _selection: SelectionInfo | undefined): PresentationRpcResponse<DescriptorJSON | undefined> { return this.forward(arguments); }
  public async getContentSetSize(_token: IModelRpcProps, _options: ContentRpcRequestOptions, _descriptorOrOverrides: DescriptorJSON | DescriptorOverrides, _keys: KeySetJSON): PresentationRpcResponse<number> { return this.forward(arguments); }
  public async getContent(_token: IModelRpcProps, _options: ContentRpcRequestOptions, _descriptorOrOverrides: DescriptorJSON | DescriptorOverrides, _keys: KeySetJSON): PresentationRpcResponse<ContentJSON | undefined> { return this.forward(arguments); }
  public async getContentAndSize(_token: IModelRpcProps, _options: ContentRpcRequestOptions, _descriptorOrOverrides: DescriptorJSON | DescriptorOverrides, _keys: KeySetJSON): PresentationRpcResponse<{ content?: ContentJSON, size: number }> { return this.forward(arguments); }
  public async getDistinctValues(_token: IModelRpcProps, _options: ContentRpcRequestOptions, _descriptor: DescriptorJSON, _keys: KeySetJSON, _fieldName: string, _maximumValueCount: number): PresentationRpcResponse<string[]> { return this.forward(arguments); }

  public async getDisplayLabelDefinition(_token: IModelRpcProps, _options: LabelRpcRequestOptions, _key: InstanceKeyJSON): PresentationRpcResponse<LabelDefinitionJSON> { return this.forward(arguments); }
  public async getDisplayLabelDefinitions(_token: IModelRpcProps, _options: LabelRpcRequestOptions, _keys: InstanceKeyJSON[]): PresentationRpcResponse<LabelDefinitionJSON[]> { return this.forward(arguments); }

  public async getSelectionScopes(_token: IModelRpcProps, _options: SelectionScopeRpcRequestOptions): PresentationRpcResponse<SelectionScope[]> { return this.forward(arguments); }
  public async computeSelection(_token: IModelRpcProps, _options: SelectionScopeRpcRequestOptions, _ids: Id64String[], _scopeId: string): PresentationRpcResponse<KeySetJSON> { return this.forward(arguments); }

  /** @alpha */
  public async compareHierarchies(_token: IModelRpcProps, _options: PresentationDataCompareRpcOptions): PresentationRpcResponse<PartialHierarchyModificationJSON[]> { return this.forward(arguments); }
}

/** @alpha */
export enum PresentationRpcEvents {
  /**
   * ID of an event that's emitted when backend detects changes in presented data.
   */
  Update = "OnUpdate",
}
