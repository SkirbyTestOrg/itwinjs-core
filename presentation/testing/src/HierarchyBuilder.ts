/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { using } from "@bentley/bentleyjs-core";
import { Ruleset, Omit, RegisteredRuleset } from "@bentley/presentation-common";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Presentation } from "@bentley/presentation-frontend";
import { PresentationTreeDataProvider } from "@bentley/presentation-components";
import { TreeNodeItem } from "@bentley/ui-components";

/** A derivative of Node. Cannot have **children** property */
export interface MappedNode {
  [index: string]: any;
  children?: never;
}

/** Node in a Hierarchy */
export interface HierarchyNode extends Omit<MappedNode, "children"> {
  children?: HierarchyNode[];
}

/** A function that converts `TreeNodeItem` into a new custom object */
export type NodeMappingFunc = (node: TreeNodeItem) => MappedNode;

/** Default [[NodeMappingFunc]] implementation that outputs the whole `TreeNodeItem` object */
export const defaultNodeMappingFunc: NodeMappingFunc = (node: TreeNodeItem) => {
  // Skip properties 'id', 'parentId', 'extendedData' as they contain  internal stuff
  // that callers are most likely not interested in. Otherwise they can supply
  // a custom `NodeMappingFunc` that does return those properties as well.
  const { id, parentId, extendedData, ...resultNode } = node;
  return resultNode;
};

/**
 * A class that constructs simple node hierarchy from specified
 * imodel and ruleset.
 */
export class HierarchyBuilder {
  private readonly _iModel: IModelConnection;
  private readonly _nodeMappingFunc: NodeMappingFunc;

  /**
   * Constructor
   * @param iModel The iModel to pull data from
   * @param nodeMappingFunc A function that maps node to something that the user of
   * this API is interested in. E.g. custom implementation may skip some unimportant
   * node properties to make resulting object smaller and easier to read.
   */
  constructor(iModel: IModelConnection, nodeMappingFunc: NodeMappingFunc = defaultNodeMappingFunc) {
    this._iModel = iModel;
    this._nodeMappingFunc = nodeMappingFunc;
  }

  private async createSubHierarchy(nodes: TreeNodeItem[], dataProvider: PresentationTreeDataProvider) {
    const hierarchy: HierarchyNode[] = [];
    for (const node of nodes) {
      const nodeIndex = hierarchy.push(this._nodeMappingFunc(node)) - 1;
      const childNodes = await dataProvider.getNodes(node);
      if (childNodes.length > 0)
        hierarchy[nodeIndex].children = await this.createSubHierarchy(childNodes, dataProvider);
    }
    return hierarchy;
  }

  private async doCreateHierarchy(rulesetId: string): Promise<HierarchyNode[]> {
    const dataProvider = new PresentationTreeDataProvider(this._iModel, rulesetId);
    const rootNodes = await dataProvider.getNodes();
    return this.createSubHierarchy(rootNodes, dataProvider);
  }

  /**
   * Create a hierarchy using the supplied presentation ruleset.
   * @param rulesetOrId Either a [[Ruleset]] object or a ruleset id.
   */
  public async createHierarchy(rulesetOrId: Ruleset | string): Promise<HierarchyNode[]> {
    if (typeof rulesetOrId === "string")
      return this.doCreateHierarchy(rulesetOrId);

    return using(await Presentation.presentation.rulesets().add(rulesetOrId), async (ruleset: RegisteredRuleset) => {
      return this.doCreateHierarchy(ruleset.id);
    });
  }
}
