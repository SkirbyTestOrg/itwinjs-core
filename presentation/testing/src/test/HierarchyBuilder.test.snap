// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchyBuilder createHierarchy with data returns correct hierarchy 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "label": "Child 1",
      },
      Object {
        "label": "Child 2",
      },
    ],
    "hasChildren": true,
    "label": "Root Node",
  },
]
`;

exports[`HierarchyBuilder createHierarchy with data returns correct hierarchy with custom node mapping function 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "id": "child1/root",
      },
      Object {
        "id": "child2/root",
      },
    ],
    "id": "root",
  },
]
`;
